### ISSUE TITLE
Upgrade to checkstyle 8.21

### ISSUE THREAD

**Issue by rnveach** [sevntu-checkstyle/sevntu.checkstyle#745]
Upgrade to checkstyle 8.21 .
Blocked by eclipse-cs.
main repo regression testing on sevntu is disabled.

Compilation failures:
Test:
````
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/PublicReferenceToPrivateTypeCheckTest.java:[385,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/EmptyPublicCtorInClassCheckTest.java:[179,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/RedundantReturnCheckTest.java:[152,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/AvoidConditionInversionCheckTest.java:[89,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/CauseParameterInExceptionCheckTest.java:[163,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[294,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[297,33] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[301,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[303,13] cannot find symbol
  symbol:   method addChild(com.puppycrawl.tools.checkstyle.api.DetailAST)
  location: variable type of type com.puppycrawl.tools.checkstyle.api.DetailAST
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[305,33] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[307,14] cannot find symbol
  symbol:   method addChild(com.puppycrawl.tools.checkstyle.api.DetailAST)
  location: variable child of type com.puppycrawl.tools.checkstyle.api.DetailAST
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[309,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[310,13] cannot find symbol
  symbol:   method addChild(com.puppycrawl.tools.checkstyle.api.DetailAST)
  location: variable node of type com.puppycrawl.tools.checkstyle.api.DetailAST
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[311,13] cannot find symbol
  symbol:   method addChild(com.puppycrawl.tools.checkstyle.api.DetailAST)
  location: variable node of type com.puppycrawl.tools.checkstyle.api.DetailAST
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[313,34] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheckTest.java:[315,15] cannot find symbol
  symbol:   method addChild(com.puppycrawl.tools.checkstyle.api.DetailAST)
  location: variable parent of type com.puppycrawl.tools.checkstyle.api.DetailAST
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/CheckstyleTestMakeupCheckTest.java:[116,34] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/ReturnNullInsteadOfBooleanCheckTest.java:[55,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/ReturnNullInsteadOfBooleanCheckTest.java:[71,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/ForbidCertainImportsCheckTest.java:[203,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/ForbidInstantiationCheckTest.java:[129,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheckTest.java:[100,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
[ERROR] /sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/ForbidCertainMethodCheckTest.java:[154,32] com.puppycrawl.tools.checkstyle.api.DetailAST is abstract; cannot be instantiated
````

**Comment by romani** [comment]
@rnveach ,

We did very big compatibility damage - https://travis-ci.org/sevntu-checkstyle/checkstyle-samples/jobs/537427424#L522 , please recheck why it was not detected during testing.

```
[checkstyle] Running Checkstyle 8.21 on 1 files
BUILD FAILED
/home/travis/build/sevntu-checkstyle/checkstyle-samples/ant-project/build.xml:15: java.lang.Error: Error was thrown while processing /home/travis/build/sevntu-checkstyle/checkstyle-samples/ant-project/src/com/github/sevntu/checkstyle/Sample.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:315)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:220)
	at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.processFiles(CheckstyleAntTask.java:349)
	at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.realExecute(CheckstyleAntTask.java:316)
	at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.execute(CheckstyleAntTask.java:288)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:280)
	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:109)
Caused by: java.lang.IncompatibleClassChangeError: Found interface com.puppycrawl.tools.checkstyle.api.DetailAST, but class was expected
	at com.github.sevntu.checkstyle.checks.coding.ReturnNullInsteadOfBooleanCheck.visitToken(ReturnNullInsteadOfBooleanCheck.java:78)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:363)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:470)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:304)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:162)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:85)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:332)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:294)
	... 21 more
```

**Comment by rnveach** [comment]
See https://github.com/checkstyle/checkstyle/issues/3417#issuecomment-496018795 .
I assume sevntu does the same testing and rebuilds the jars from scratch instead of using the released versions.

**Comment by romani** [comment]
Sevntu samples use only release version https://github.com/sevntu-checkstyle/checkstyle-samples/blob/master/.travis.yml
https://github.com/sevntu-checkstyle/checkstyle-samples/blob/master/maven-project/pom.xml#L14

**Comment by pzygielo** [comment]
Hi team, as it's two months now since last activity here, I just wonder - is there a plan to release version compatible with current CS, or somehow sync them?
I'm stuck at CS-8.20, is that the only option for now?

**Comment by romani** [comment]
We all stuck with old eclipse-cs, but we need somebody to help us to update eclipse-cs project to most recent checkstyle.
Or figure out way how to split sevntu project to detach dependant to eclipse-cs to only for eclipse-cs extension. Feel free to suggest solution and implementation.

**Comment by pzygielo** [comment]
> Feel free to suggest solution and implementation.

Funny :smile: <sub>(But should my CS/sevntu studies result in anything of value I'll share.)</sub>

> We all stuck with old eclipse-cs, 

For now thanks for confirming it's not only me/my configuration.

**Comment by romani** [comment]
>But should my CS/sevntu studies result in anything of value I'll share.

We are open for suggestions, there are a lot of rooms to improve. Unfortunately our tiny team is focused on main project, and even there we need help.

**Comment by romani** [comment]
fixed in scope of #782
